script/testing/self_driving/forecast.py

```diff
@@ -54,73 +52,200 @@ def config_forecast_data(xml_config_file: str, rate_pattern: List[int]) -> None:
     xml.write(xml_config_file)
 
 
-def gen_oltp_trace(
-        tpcc_weight: str, tpcc_rates: List[int], pattern_iter: int, record_pipeline_metrics: bool) -> bool:
+def get_config_tpcc(weights=constants.DEFAULT_TPCC_WEIGHTS,
+                    terminals=constants.DEFAULT_TPCC_TERMINALS,
+                    scale_factor=constants.DEFAULT_TPCC_SCALE_FACTOR,
+                    query_mode=constants.DEFAULT_TPCC_QUERY_MODE) -> dict:
     """
-    Generate the trace by running OLTPBench's TPCC benchmark on the built DBMS.
+    Get a configuration for running TPC-C.
+
+    Parameters
+    ----------
+    weights : str
+        The weight for the txn types in TPC-C.
+        NewOrder, Payment, OrderStatus, Delivery, StockLevel.
+    terminals : int
+        The number of terminals to invoke in TPC-C.
+    scale_factor : int
+        The number of warehouses in TPC-C.
+    query_mode : str
+        Either "simple" or "extended"
+
+    Returns
+    -------
+    config : dict
+        A dictionary with all the relevant TPC-C parameters configured, usable as input to TestCaseOLTPBench.
+    """
+    config = {}
+    config['benchmark'] = "tpcc"
+    config['weights'] = weights
+    config['terminals'] = terminals
+    config['scale_factor'] = scale_factor
+    config['query_mode'] = query_mode
+    return config
+
+
+def fn_pipeline_metrics(oltpbench: TestOLTPBench, test_case: TestCaseOLTPBench) -> None:
+    """Enable pipeline metrics."""
+    db_server = oltpbench.db_instance
+    db_server.execute("SET pipeline_metrics_enable='true'", expect_result=False, quiet=False)
+    db_server.execute("SET pipeline_metrics_sample_rate={}".format(constants.DEFAULT_PIPELINE_METRICS_SAMPLE_RATE),
+                      expect_result=False, quiet=False)
+
+
+def fn_query_trace_metrics(oltpbench: TestOLTPBench, test_case: TestCaseOLTPBench) -> None:
+    """Enable query trace metrics."""
+    db_server = oltpbench.db_instance
+    db_server.execute("SET query_trace_metrics_enable='true'", expect_result=False, quiet=False)
+
+
+def fn_enable_pilot(oltpbench: TestOLTPBench, test_case: TestCaseOLTPBench) -> None:
+    """Enable pilot planning."""
+    db_server = oltpbench.db_instance
+    db_server.execute("SET pilot_planning='true'", expect_result=False, quiet=False)
+
+
+def fn_test_results_file(result_file_path: str) -> bool:
+    """Test if the specified results file exists (DEFAULT_PIPELINE_METRICS_FILE or DEFAULT_QUERY_TRACE_FILE."""
+    if not Path(result_file_path).exists():
+        LOG.error(f"Missing {result_file_path} at CWD.")
+        return False
+    return True
+
+
+def make_fn_test_case_run_pre_test(create: bool, load: bool) -> Callable[[TestOLTPBench, TestCaseOLTPBench], None]:
```

Comment:
Just curious why do you choose this functor design? What property do you want to achieve?

Reply:
As mentioned in person the other day, but here for posterity:

It is unlikely that the script will always fit in all the functionality that we want. As we have changing requirements and different things that we want to add in the future, we have the following options:

- Modify the function being called into (then you must understand all the current callers and/or gate it behind additional flags)
- Duplicate the function being called into (then you have code duplication)

If it is a functor based design where the caller provides all the things that should happen, then it should be easier to extend functionality in the future (just add a new functor and have the caller invoke it, all existing call sites remain the same).